src/main/scala/io/iohk/ethereum/network/rlpx/RLPxConnectionHandler.scala

```diff
@@ -136,21 +140,64 @@ class RLPxConnectionHandler(
         context.parent ! MessageReceived(message)
 
       case Failure(ex) =>
-        log.error(ex, "Cannot decode message")
+        log.error(ex, s"Cannot decode message from $peerId")
     }
 
-    def handshaked(messageCodec: MessageCodec): Receive =
+    /**
+      * Handles sending and receiving messages from the Akka TCP connection, while also handling acknowledgement of
+      * messages sent. Messages are only sent when all Ack from previous messages were received.
+      *
+      * @param messageCodec, for encoding the messages sent
+      * @param messagesNotSent, messages not yet sent
+      * @param tcpAckTimeout, timeout for the message sent for which we are awaiting an acknowledgement (if there is one)
+      */
+    def handshaked(messageCodec: MessageCodec,
+                   messagesNotSent: Queue[MessageSerializable] = Queue.empty,
+                   tcpAckTimeout: Option[Cancellable] = None): Receive =
       handleWriteFailed orElse handleConnectionClosed orElse {
         case sm: SendMessage =>
-          val out = messageCodec.encodeMessage(sm.serializable)
-          connection ! Write(out)
-          log.debug("Sent message: {}", sm.serializable)
+          if(tcpAckTimeout.isEmpty) {
+            sendMessage(messageCodec, sm.serializable)
+            val timeout = system.scheduler.scheduleOnce(rlpxConfiguration.waitForTcpAckTimeout, self, AckTimeout)
+            context become handshaked(messageCodec, messagesNotSent, Some(timeout))
+          } else
+            context become handshaked(messageCodec, messagesNotSent :+ sm.serializable, tcpAckTimeout)
 
         case Received(data) =>
           val messages = messageCodec.readMessages(data)
           messages foreach processMessage
+
+        case Ack if tcpAckTimeout.nonEmpty =>
+          //Cancel pending message timeout
+          tcpAckTimeout.foreach(_.cancel())
```

Comment:
is it possible that we get race condition where timeout will send AckTimeout message to us before we cancel it and as a result this timeout will be treated as timeout for message we are sending in this case
maybe we can distinguish AckTimeout somehow to be able to determine which AckTimeout is for which message?